File: packages/mcp-servers/waveform-mcp/src/waveform_mcp/skills/waveform_debug/detectors.py
"""Five parallel WaveformDebug detectors.

Design principle: Programs do detection (scanning hundreds of thousands of
time steps, precisely calculating edge widths, protocol compliance).
LLM does reasoning (understanding race hazard principles, repair solutions).

Detectors:
  1. GlitchDetector       — pulses narrower than 1 clock period
  2. XZStateTracker       — X/Z propagation tracking
  3. CDCDetector          — cross-clock-domain signal sampling hazards
  4. AXIProtocolDecoder   — AXI4 handshake violations (via axi_decoder)
  5. HandshakeTimeout     — VALID asserted without READY for > N clocks
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field


@dataclass
class Anomaly:
    detector: str
    severity: str           # "error" | "warning" | "info"
    signal:   str
    time_ns:  float
    anomaly_type: str
    message:  str
    rtl_hint: str = ""      # hint for RTL reverse mapping
    context:  dict = field(default_factory=dict)


# ── 1. Glitch Detector ────────────────────────────────────────────────────────

def detect_glitches(
    signal_events: dict[str, list[dict]],   # name → [{time, value}]
    clock_period_ns: float,
) -> list[Anomaly]:
    """Find pulses narrower than one clock period.

    Distinguishes race hazards (combinational glitches) from real short pulses.
    A glitch is defined as: value changes A→B→A within < clock_period_ns.
    """
    anomalies: list[Anomaly] = []

    if clock_period_ns <= 0:
        return anomalies

    for name, events in signal_events.items():
        if len(events) < 3:
            continue
        for i in range(1, len(events) - 1):
            t0, v0 = events[i - 1]["time"], events[i - 1]["value"]
            t1, v1 = events[i    ]["time"], events[i    ]["value"]
            t2, v2 = events[i + 1]["time"], events[i + 1]["value"]

            if v1 != v0 and v1 != v2 and v0 == v2:
                pulse_width = t2 - t1
                if 0 < pulse_width < clock_period_ns:
                    anomalies.append(Anomaly(
                        detector="GlitchDetector",
                        severity="warning",
                        signal=name,
                        time_ns=t1,
                        anomaly_type="glitch",
                        message=(
                            f"Glitch on {name}: pulse width {pulse_width:.2f} ns "
                            f"< 1 clock period ({clock_period_ns:.2f} ns). "
                            f"Possible combinational race hazard."
                        ),
                        rtl_hint=f"Check combinational logic driving {name}",
                    ))
    return anomalies


# ── 2. X/Z State Tracker ─────────────────────────────────────────────────────

def track_xz_states(
    signal_events: dict[str, list[dict]],
) -> list[Anomaly]:
    """Record signals entering X or Z states and flag them."""
    anomalies: list[Anomaly] = []

    for name, events in signal_events.items():
        in_xz = False
        entry_time: float = 0.0

        for evt in events:
            val = str(evt["value"]).lower()
            is_xz = any(c in val for c in ("x", "z"))

            if is_xz and not in_xz:
                in_xz = True
                entry_time = evt["time"]
                anomalies.append(Anomaly(
                    detector="XZStateTracker",
                    severity="error",
                    signal=name,
                    time_ns=entry_time,
                    anomaly_type="x_state" if "x" in val else "z_state",
                    message=(
                        f"{name} entered {'X' if 'x' in val else 'Z'} state at "
                        f"{entry_time:.1f} ns. Possible uninitialized register or "
                        f"missing reset."
                    ),
                    rtl_hint=f"Check reset coverage for {name}; verify no undriven nets",
                ))
            elif not is_xz:
                in_xz = False

    return anomalies


# ── 3. CDC Detector ───────────────────────────────────────────────────────────

def detect_cdc(
    signal_events: dict[str, list[dict]],
    clock_names: list[str],
    min_sync_stages: int = 2,
) -> list[Anomaly]:
    """Identify signals that transition near multiple different clock edges.

    Heuristic: a signal that has transitions within half a cycle of two
    different clocks is a potential CDC crossing without synchronization.

    Note: Full CDC analysis requires RTL structural analysis (eda-bridge-mcp).
    This detector provides timing-domain evidence.
    """
    anomalies: list[Anomaly] = []

    if len(clock_names) < 2:
        return anomalies

    # Collect rising edges for each clock
    clock_edges: dict[str, list[float]] = {}
    for clk_name in clock_names:
        evts = signal_events.get(clk_name, [])
        edges: list[float] = []
        for i in range(1, len(evts)):
            if evts[i - 1]["value"] in ("0", "x") and evts[i]["value"] == "1":
                edges.append(evts[i]["time"])
        clock_edges[clk_name] = edges

    # Estimate clock periods
    periods: dict[str, float] = {}
    for clk, edges in clock_edges.items():
        if len(edges) >= 2:
            periods[clk] = edges[1] - edges[0]

    # For each non-clock signal, check proximity to multiple clocks
    for name, events in signal_events.items():
        if name in clock_names:
            continue

        transitions = [e["time"] for i, e in enumerate(events[1:], 1)
                       if e["value"] != events[i - 1]["value"]]

        clocks_nearby: dict[str, int] = {c: 0 for c in clock_names}

        for t in transitions:
            for clk, edges in clock_edges.items():
                period = periods.get(clk, 10.0)
                tolerance = period * 0.1  # 10% of period
                near = any(abs(t - edge) < tolerance for edge in edges)
                if near:
                    clocks_nearby[clk] += 1

        active_clocks = [c for c, cnt in clocks_nearby.items() if cnt > 0]
        if len(active_clocks) >= 2:
            anomalies.append(Anomaly(
                detector="CDCDetector",
                severity="warning",
                signal=name,
                time_ns=transitions[0] if transitions else 0.0,
                anomaly_type="cdc_suspect",
                message=(
                    f"{name} transitions correlate with {len(active_clocks)} clock domains "
                    f"({', '.join(active_clocks)}). Possible CDC crossing — verify "
                    f"{min_sync_stages}-FF synchronizer is present."
                ),
                rtl_hint=f"Search RTL for {name} to verify 2-FF synchronizer",
            ))

    return anomalies


# ── 4. AXI Protocol Decoder (via axi_decoder module) ─────────────────────────

def detect_axi_violations(
    signal_events: dict[str, list[dict]],
    axi_prefix: str = "",
    clock_name: str | None = None,
) -> list[Anomaly]:
    """Wrapper around axi_decoder — converts violations to Anomaly objects."""
    try:
        from waveform_mcp.axi_decoder import decode_axi
    except ImportError:
        # Running outside waveform-mcp context; skip silently
        return []

    result = decode_axi(signal_events, axi_prefix=axi_prefix, clock_name=clock_name)
    anomalies: list[Anomaly] = []

    for v in result.get("violations", []):
        anomalies.append(Anomaly(
            detector="AXIProtocolDecoder",
            severity="error",
            signal=f"{axi_prefix}{v['channel'].lower()}valid",
            time_ns=v["time_ns"],
            anomaly_type=v["violation_type"],
            message=v["message"],
            rtl_hint=v["suggestion"],
        ))

    return anomalies


# ── 5. Handshake Timeout Detector ─────────────────────────────────────────────

def detect_handshake_timeouts(
    signal_events: dict[str, list[dict]],
    clock_period_ns: float,
    timeout_cycles: int = 100,
    valid_signal_patterns: list[str] | None = None,
) -> list[Anomaly]:
    """Detect valid/req signals asserted without corresponding ack/ready.

    Default patterns: signals ending in _valid, _req, _en.
    """
    patterns = valid_signal_patterns or ["_valid", "_req", "_en", "valid", "req"]
    timeout_ns = timeout_cycles * clock_period_ns
    anomalies: list[Anomaly] = []

    for name, events in signal_events.items():
        # Only check signals that look like valid/req
        if not any(pat in name for pat in patterns):
            continue

        # Check for long assertion
        assert_start: float | None = None
        for evt in events:
            val = evt["value"]
            t   = evt["time"]

            if val == "1":
                if assert_start is None:
                    assert_start = t
                elif t - assert_start > timeout_ns:
                    anomalies.append(Anomaly(
                        detector="HandshakeTimeout",
                        severity="warning",
                        signal=name,
                        time_ns=assert_start,
                        anomaly_type="handshake_timeout",
                        message=(
                            f"{name} asserted for >{timeout_cycles} clocks "
                            f"({t - assert_start:.0f} ns) without deassertion. "
                            f"Check if partner ready/ack signal responds."
                        ),
                        rtl_hint=f"Inspect downstream block connected to {name}",
                    ))
                    assert_start = None  # avoid duplicate reports
            else:
                assert_start = None

    return anomalies


# ── Run all detectors in parallel ─────────────────────────────────────────────

async def run_all_detectors(
    signal_events: dict[str, list[dict]],
    clock_period_ns: float = 10.0,
    clock_names: list[str] | None = None,
    axi_prefix: str = "",
    timeout_cycles: int = 100,
) -> list[Anomaly]:
    """Run all 5 detectors concurrently and merge results."""
    loop = asyncio.get_event_loop()

    clk_names = clock_names or []

    tasks = [
        loop.run_in_executor(None, detect_glitches,            signal_events, clock_period_ns),
        loop.run_in_executor(None, track_xz_states,            signal_events),
        loop.run_in_executor(None, detect_cdc,                 signal_events, clk_names),
        loop.run_in_executor(None, detect_axi_violations,      signal_events, axi_prefix, None),
        loop.run_in_executor(None, detect_handshake_timeouts,  signal_events, clock_period_ns, timeout_cycles),
    ]

    results = await asyncio.gather(*tasks, return_exceptions=True)
    anomalies: list[Anomaly] = []

    for r in results:
        if isinstance(r, list):
            anomalies.extend(r)
        # Silently skip exceptions from individual detectors

    # Sort by severity (error first) then by time
    severity_order = {"error": 0, "warning": 1, "info": 2}
    anomalies.sort(key=lambda a: (severity_order.get(a.severity, 3), a.time_ns))

    return anomalies

File: packages/mcp-servers/waveform-mcp-rs/src/detectors.rs
//! Five waveform anomaly detectors.
//!
//! Ports `detectors.py` from the deleted Python `waveform-mcp` package. Each
//! detector returns a `Vec<Anomaly>`. `run_all` executes them concurrently on
//! a tokio runtime via `spawn_blocking` (each detector is pure CPU).
//!
//! Carries the bug fixes we made on the Python side last round:
//!   * Glitch detector skips clock signals (bug #12).
//!   * CDC detector deduplicates clock aliases by edge fingerprint (bug #13).

use serde::Serialize;
use serde_json::{json, Map, Value};

#[derive(Clone, Debug, Serialize)]
pub struct Anomaly {
    pub detector: String,
    pub severity: String,        // "error" | "warning" | "info"
    pub signal:   String,
    pub time_ns:  f64,
    pub anomaly_type: String,
    pub message:  String,
    pub rtl_hint: String,
}

impl Anomaly {
    pub fn to_json_with_index(&self, index: usize) -> Value {
        json!({
            "index":    index,
            "detector": self.detector,
            "severity": self.severity,
            "signal":   self.signal,
            "time_ns":  self.time_ns,
            "type":     self.anomaly_type,
            "message":  self.message,
            "rtl_hint": self.rtl_hint,
        })
    }
}

/// An event payload as produced by the compressor: `{time: f64, value: str, [stable: bool]}`.
/// Helper accessors for the detector code.
fn event_time(ev: &Value) -> f64 {
    ev.get("time").and_then(|v| v.as_f64()).unwrap_or(0.0)
}

fn event_value(ev: &Value) -> &str {
    ev.get("value").and_then(|v| v.as_str()).unwrap_or("")
}

// ── 1. Glitch Detector ───────────────────────────────────────────────────────

pub fn detect_glitches(
    signal_events: &Map<String, Value>,
    clock_period_ns: f64,
    clock_names: &[String],
) -> Vec<Anomaly> {
    let mut anomalies = Vec::new();
    if clock_period_ns <= 0.0 {
        return anomalies;
    }

    let skip: std::collections::HashSet<&String> = clock_names.iter().collect();

    for (name, events_val) in signal_events {
        if skip.contains(name) {
            continue;
        }
        let events = match events_val.as_array() {
            Some(arr) if arr.len() >= 3 => arr,
            _ => continue,
        };

        for i in 1..events.len() - 1 {
            let v0 = event_value(&events[i - 1]);
            let v1 = event_value(&events[i]);
            let v2 = event_value(&events[i + 1]);
            if v1 != v0 && v1 != v2 && v0 == v2 {
                let t1 = event_time(&events[i]);
                let t2 = event_time(&events[i + 1]);
                let pulse_width = t2 - t1;
                if pulse_width > 0.0 && pulse_width < clock_period_ns {
                    anomalies.push(Anomaly {
                        detector: "GlitchDetector".into(),
                        severity: "warning".into(),
                        signal: name.clone(),
                        time_ns: t1,
                        anomaly_type: "glitch".into(),
                        message: format!(
                            "Glitch on {}: pulse width {:.2} ns < 1 clock period ({:.2} ns). \
                             Possible combinational race hazard.",
                            name, pulse_width, clock_period_ns
                        ),
                        rtl_hint: format!("Check combinational logic driving {}", name),
                    });
                }
            }
        }
    }
    anomalies
}

// ── 2. X/Z State Tracker ─────────────────────────────────────────────────────

pub fn track_xz_states(signal_events: &Map<String, Value>) -> Vec<Anomaly> {
    let mut anomalies = Vec::new();
    for (name, events_val) in signal_events {
        let events = match events_val.as_array() {
            Some(a) => a,
            None => continue,
        };
        let mut in_xz = false;
        for evt in events {
            let val = event_value(evt).to_ascii_lowercase();
            let is_xz = val.contains('x') || val.contains('z');
            if is_xz && !in_xz {
                in_xz = true;
                let kind = if val.contains('x') { 'X' } else { 'Z' };
                let t = event_time(evt);
                anomalies.push(Anomaly {
                    detector: "XZStateTracker".into(),
                    severity: "error".into(),
                    signal: name.clone(),
                    time_ns: t,
                    anomaly_type: if kind == 'X' { "x_state" } else { "z_state" }.into(),
                    message: format!(
                        "{} entered {} state at {:.1} ns. Possible uninitialized register or \
                         missing reset.",
                        name, kind, t
                    ),
                    rtl_hint: format!("Check reset coverage for {}; verify no undriven nets", name),
                });
            } else if !is_xz {
                in_xz = false;
            }
        }
    }
    anomalies
}

// ── 3. CDC Detector ──────────────────────────────────────────────────────────

pub fn detect_cdc(
    signal_events: &Map<String, Value>,
    clock_names: &[String],
    min_sync_stages: usize,
) -> Vec<Anomaly> {
    let mut anomalies = Vec::new();
    if clock_names.len() < 2 {
        return anomalies;
    }

    // Collect rising edges for each clock.
    let mut clock_edges: Vec<(String, Vec<f64>)> = Vec::new();
    for clk in clock_names {
        let events = signal_events
            .get(clk)
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        let mut edges = Vec::new();
        for i in 1..events.len() {
            let prev = event_value(&events[i - 1]);
            let cur = event_value(&events[i]);
            if (prev == "0" || prev == "x") && cur == "1" {
                edges.push(event_time(&events[i]));
            }
        }
        clock_edges.push((clk.clone(), edges));
    }

    // Dedup clock aliases by fingerprint of the first 6 edges (bug #13).
    let mut unique: Vec<(String, Vec<f64>)> = Vec::new();
    for (name, edges) in clock_edges.into_iter() {
        if edges.is_empty() {
            continue;
        }
        let fp: &[f64] = &edges[..edges.len().min(6)];
        let already = unique.iter().any(|(_, e)| {
            let other_fp: &[f64] = &e[..e.len().min(6)];
            fp.len() == other_fp.len()
                && fp
                    .iter()
                    .zip(other_fp.iter())
                    .all(|(a, b)| (a - b).abs() < 0.001)
        });
        if !already {
            unique.push((name, edges));
        }
    }
    if unique.len() < 2 {
        return anomalies;
    }

    // Estimate per-clock period.
    let periods: Vec<f64> = unique
        .iter()
        .map(|(_, edges)| if edges.len() >= 2 { edges[1] - edges[0] } else { 10.0 })
        .collect();

    let clock_name_set: std::collections::HashSet<&String> =
        clock_names.iter().collect();

    for (name, events_val) in signal_events {
        if clock_name_set.contains(name) {
            continue;
        }
        let events = match events_val.as_array() {
            Some(a) => a,
            None => continue,
        };

        let mut transitions = Vec::new();
        for i in 1..events.len() {
            if event_value(&events[i]) != event_value(&events[i - 1]) {
                transitions.push(event_time(&events[i]));
            }
        }

        let mut counts = vec![0usize; unique.len()];
        for &t in &transitions {
            for (idx, ((_clk, edges), &period)) in unique.iter().zip(periods.iter()).enumerate() {
                let tol = period * 0.1;
                if edges.iter().any(|&e| (t - e).abs() < tol) {
                    counts[idx] += 1;
                }
            }
        }

        let active: Vec<&str> = unique
            .iter()
            .zip(counts.iter())
            .filter_map(|((clk, _), &c)| if c > 0 { Some(clk.as_str()) } else { None })
            .collect();

        if active.len() >= 2 {
            anomalies.push(Anomaly {
                detector: "CDCDetector".into(),
                severity: "warning".into(),
                signal: name.clone(),
                time_ns: *transitions.first().unwrap_or(&0.0),
                anomaly_type: "cdc_suspect".into(),
                message: format!(
                    "{} transitions correlate with {} clock domains ({}). Possible CDC \
                     crossing — verify {}-FF synchronizer is present.",
                    name,
                    active.len(),
                    active.join(", "),
                    min_sync_stages
                ),
                rtl_hint: format!("Search RTL for {} to verify 2-FF synchronizer", name),
            });
        }
    }

    anomalies
}

// ── 4. AXI Protocol Violations (wraps axi_decoder) ───────────────────────────

pub fn detect_axi_violations(
    signal_events: &Map<String, Value>,
    axi_prefix: &str,
    clock_name: Option<&str>,
) -> Vec<Anomaly> {
    if axi_prefix.is_empty() {
        return Vec::new();
    }
    let result = crate::axi_decoder::decode_axi(signal_events, axi_prefix, clock_name, 100);
    let mut anomalies = Vec::new();
    if let Some(violations) = result.get("violations").and_then(|v| v.as_array()) {
        for v in violations {
            let channel = v.get("channel").and_then(|x| x.as_str()).unwrap_or("?");
            anomalies.push(Anomaly {
                detector: "AXIProtocolDecoder".into(),
                severity: "error".into(),
                signal: format!("{}{}valid", axi_prefix, channel.to_lowercase()),
                time_ns: v.get("time_ns").and_then(|x| x.as_f64()).unwrap_or(0.0),
                anomaly_type: v
                    .get("violation_type")
                    .and_then(|x| x.as_str())
                    .unwrap_or("axi_violation")
                    .to_string(),
                message: v
                    .get("message")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
                rtl_hint: v
                    .get("suggestion")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
            });
        }
    }
    anomalies
}

// ── 5. Handshake Timeout Detector ────────────────────────────────────────────

pub fn detect_handshake_timeouts(
    signal_events: &Map<String, Value>,
    clock_period_ns: f64,
    timeout_cycles: u32,
) -> Vec<Anomaly> {
    let patterns = ["_valid", "_req", "_en", "valid", "req"];
    let timeout_ns = timeout_cycles as f64 * clock_period_ns;
    let mut anomalies = Vec::new();

    for (name, events_val) in signal_events {
        if !patterns.iter().any(|p| name.contains(p)) {
            continue;
        }
        let events = match events_val.as_array() {
            Some(a) => a,
            None => continue,
        };
        let mut assert_start: Option<f64> = None;
        for evt in events {
            let val = event_value(evt);
            let t = event_time(evt);
            if val == "1" {
                if assert_start.is_none() {
                    assert_start = Some(t);
                } else if let Some(start) = assert_start {
                    if t - start > timeout_ns {
                        anomalies.push(Anomaly {
                            detector: "HandshakeTimeout".into(),
                            severity: "warning".into(),
                            signal: name.clone(),
                            time_ns: start,
                            anomaly_type: "handshake_timeout".into(),
                            message: format!(
                                "{} asserted for >{} clocks ({:.0} ns) without deassertion. \
                                 Check if partner ready/ack signal responds.",
                                name,
                                timeout_cycles,
                                t - start
                            ),
                            rtl_hint: format!("Inspect downstream block connected to {}", name),
                        });
                        assert_start = None;
                    }
                }
            } else {
                assert_start = None;
            }
        }
    }
    anomalies
}

// ── Run all detectors in parallel ────────────────────────────────────────────

pub async fn run_all_detectors(
    signal_events: Map<String, Value>,
    clock_period_ns: f64,
    clock_names: Vec<String>,
    axi_prefix: String,
    timeout_cycles: u32,
) -> Vec<Anomaly> {
    let events = std::sync::Arc::new(signal_events);
    let clocks = std::sync::Arc::new(clock_names);
    let prefix = std::sync::Arc::new(axi_prefix);

    let e1 = events.clone();
    let c1 = clocks.clone();
    let t1 = tokio::task::spawn_blocking(move || detect_glitches(&e1, clock_period_ns, &c1));

    let e2 = events.clone();
    let t2 = tokio::task::spawn_blocking(move || track_xz_states(&e2));

    let e3 = events.clone();
    let c3 = clocks.clone();
    let t3 = tokio::task::spawn_blocking(move || detect_cdc(&e3, &c3, 2));

    let e4 = events.clone();
    let p4 = prefix.clone();
    let t4 = tokio::task::spawn_blocking(move || detect_axi_violations(&e4, &p4, None));

    let e5 = events.clone();
    let t5 = tokio::task::spawn_blocking(move || {
        detect_handshake_timeouts(&e5, clock_period_ns, timeout_cycles)
    });

    let mut anomalies: Vec<Anomaly> = Vec::new();
    for join in [t1, t2, t3, t4, t5] {
        if let Ok(part) = join.await {
            anomalies.extend(part);
        }
    }

    // Sort by severity (error first) then by time.
    let severity_rank = |s: &str| match s {
        "error" => 0,
        "warning" => 1,
        "info" => 2,
        _ => 3,
    };
    anomalies.sort_by(|a, b| {
        severity_rank(&a.severity)
            .cmp(&severity_rank(&b.severity))
            .then(a.time_ns.partial_cmp(&b.time_ns).unwrap_or(std::cmp::Ordering::Equal))
    });
    anomalies
}

// ── Tests ────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn evt(t: f64, v: &str) -> Value {
        json!({"time": t, "value": v})
    }

    fn make_events(name: &str, points: Vec<(f64, &str)>) -> Map<String, Value> {
        let mut m = Map::new();
        m.insert(
            name.to_string(),
            Value::Array(points.into_iter().map(|(t, v)| evt(t, v)).collect()),
        );
        m
    }

    #[test]
    fn glitch_detector_skips_clock_signals() {
        // Clock with 5ns/5ns toggle would otherwise be flagged on every cycle.
        let mut m = Map::new();
        m.insert(
            "clk".to_string(),
            Value::Array(vec![
                evt(0.0, "0"),
                evt(5.0, "1"),
                evt(10.0, "0"),
                evt(15.0, "1"),
                evt(20.0, "0"),
            ]),
        );
        let anomalies = detect_glitches(&m, 10.0, &["clk".to_string()]);
        assert!(anomalies.is_empty(), "clock was misreported as glitch: {:?}", anomalies);
    }

    #[test]
    fn glitch_detector_flags_real_glitch() {
        let m = make_events(
            "data",
            vec![
                (0.0, "0"),
                (10.0, "1"),
                (12.0, "0"),  // 2ns pulse @ 10ns clock period → glitch
                (100.0, "0"),
            ],
        );
        let anomalies = detect_glitches(&m, 10.0, &[]);
        assert_eq!(anomalies.len(), 1);
        assert_eq!(anomalies[0].anomaly_type, "glitch");
    }

    #[test]
    fn xz_tracker_flags_x_entry() {
        let m = make_events("d", vec![(0.0, "0"), (5.0, "x"), (10.0, "1")]);
        let anomalies = track_xz_states(&m);
        assert_eq!(anomalies.len(), 1);
        assert_eq!(anomalies[0].anomaly_type, "x_state");
        assert_eq!(anomalies[0].severity, "error");
    }

    #[test]
    fn cdc_detector_dedups_aliased_clocks() {
        // Two clock names with identical edge timestamps — same physical net.
        let mut m = Map::new();
        let edges = vec![
            evt(0.0, "0"),
            evt(5.0, "1"),
            evt(10.0, "0"),
            evt(15.0, "1"),
            evt(20.0, "0"),
        ];
        m.insert("tb.clk".into(), Value::Array(edges.clone()));
        m.insert("tb.dut.clk".into(), Value::Array(edges));
        m.insert(
            "data".into(),
            Value::Array(vec![evt(0.0, "0"), evt(15.0, "1")]),
        );
        let anomalies = detect_cdc(
            &m,
            &["tb.clk".into(), "tb.dut.clk".into()],
            2,
        );
        assert!(anomalies.is_empty(), "alias dedup failed: {:?}", anomalies);
    }

    #[test]
    fn handshake_timeout_fires_on_long_assert() {
        let m = make_events(
            "req_valid",
            vec![(0.0, "0"), (10.0, "1"), (5000.0, "1"), (5010.0, "0")],
        );
        let anomalies = detect_handshake_timeouts(&m, 10.0, 100);
        assert_eq!(anomalies.len(), 1);
        assert_eq!(anomalies[0].anomaly_type, "handshake_timeout");
    }
}
